annotation/tests/test_tasks_services.py

```diff
@@ -580,8 +789,1416 @@ def test_filter_tasks_db_no_files_or_jobs(
+    mock_task_metric: Mock,
+    mock_agreement_score_comparing_result: Mock,
+):
+    min_match = 0.8
+    mock_parse_obj_as.return_value = [MagicMock(spec=ResponseScore)]
+    mock_task_metric.side_effect = (
+        lambda task_from_id, task_to_id, metric_score: MagicMock(
+            spec=TaskMetric,
+            task_from_id=task_from_id,
+            task_to_id=task_to_id,
+            metric_score=metric_score,
+        )
+    )
+    mock_agreement_score_comparing_result.return_value = MagicMock(
+        spec=AgreementScoreComparingResult,
+        agreement_score_reached=True,
+        task_metrics=[
+            mock_task_metric(1, 2, 0.95),
+            mock_task_metric(1, 3, 0.8),
+            mock_task_metric(2, 3, 0.85),
+        ],
+    )
+    result = services.compare_agreement_scores(
+        mock_agreement_score_response, min_match
+    )
+
+    assert result.agreement_score_reached
+
+
+def test_compare_agreement_scores_some_below_min_match(
+    mock_agreement_score_response: Mock,
+    mock_parse_obj_as: Mock,
+    mock_get_unique_scores: Mock,
+    mock_task_metric: Mock,
+    mock_agreement_score_comparing_result: Mock,
+):
+    min_match = 0.9
+    mock_parse_obj_as.return_value = [MagicMock(spec=ResponseScore)]
+    mock_task_metric.side_effect = (
+        lambda task_from_id, task_to_id, metric_score: MagicMock(
+            spec=TaskMetric,
+            task_from_id=task_from_id,
+            task_to_id=task_to_id,
+            metric_score=metric_score,
+        )
+    )
+    mock_agreement_score_comparing_result.return_value = MagicMock(
+        spec=AgreementScoreComparingResult,
+        agreement_score_reached=False,
+        task_metrics=[mock_task_metric(1, 2, 0.95)],
+    )
+    result = services.compare_agreement_scores(
+        mock_agreement_score_response, min_match
+    )
+
+    assert not result.agreement_score_reached
+
+
+def test_compare_agreement_scores_empty_response(
+    mock_parse_obj_as: Mock,
+    mock_get_unique_scores: Mock,
+    mock_task_metric: Mock,
+    mock_agreement_score_comparing_result: Mock,
+):
+    min_match = 0.5
+    mock_parse_obj_as.return_value = []
+    mock_task_metric.return_value = MagicMock(spec=TaskMetric)
+    mock_agreement_score_comparing_result.return_value = MagicMock(
+        spec=AgreementScoreComparingResult,
+        agreement_score_reached=False,
+        task_metrics=[],
+    )
+    result = services.compare_agreement_scores([], min_match)
+
+    assert not result.agreement_score_reached
+    assert result.task_metrics == []
+
+
+def test_save_agreement_metrics(mock_session: Mock):
+    agreement_score = AgreementScoreComparingResult(
+        agreement_score_reached=True, task_metrics=[]
+    )
+    services.save_agreement_metrics(mock_session, agreement_score)
+
+    mock_session.bulk_save_objects.assert_called_once_with([])
+    mock_session.commit.assert_called_once()
+
+
+def test_get_annotation_tasks(mock_session: Mock):
```

Comment:
not sure what is being tested here. it appears as input data doesnt affect the outcome of the result at all - mocks are being tested

Reply:
When I try this:


```
def test_get_annotation_tasks(mock_session: Mock):
    expected_tasks = [MagicMock(), MagicMock()]
    mock_query = mock_session.query.return_value.filter.return_value
    mock_query.all.return_value = expected_tasks

    result = services.get_annotation_tasks(
        mock_session, job_id=1, file_id=2, pages=[1, 2, 3]
    )
    mock_session.query.assert_called_once_with(ManualAnnotationTask)
    mock_session.query.return_value.filter.assert_called_once_with(
        and_(
            ManualAnnotationTask.job_id == 1,
            ManualAnnotationTask.is_validation.is_(False),
            ManualAnnotationTask.status == TaskStatusEnumSchema.finished,
            ManualAnnotationTask.file_id == 2,
            ManualAnnotationTask.pages.contained_by([1, 2, 3]),
        )
    )
    mock_query.all.assert_called_once()
    assert result == expected_tasks
```
 it prints:
```

==================================================================== FAILURES ==================================================================== 
___________________________________________________________ test_get_annotation_tasks ____________________________________________________________ 

mock_session = <NonCallableMagicMock name='Session()' spec='Session' id='2320389926624'>

    def test_get_annotation_tasks(mock_session: Mock):
        expected_tasks = [MagicMock(), MagicMock()]
        mock_query = mock_session.query.return_value.filter.return_value
        mock_query.all.return_value = expected_tasks

        result = services.get_annotation_tasks(
            mock_session, job_id=1, file_id=2, pages=[1, 2, 3]
        )

        mock_session.query.assert_called_once_with(ManualAnnotationTask)
        mock_session.query.return_value.filter.assert_called_once_with(
>           and_(
                ManualAnnotationTask.job_id == 1,
                ManualAnnotationTask.is_validation.is_(False),
                ManualAnnotationTask.status == TaskStatusEnumSchema.finished,
                ManualAnnotationTask.file_id == 2,
                ManualAnnotationTask.pages.contained_by([1, 2, 3]),
            )
        )
E       TypeError: and_ expected 2 arguments, got 5

tests\test_tasks_services.py:1352: TypeError
============================================================ short test summary info ============================================================= 
FAILED tests/test_tasks_services.py::test_get_annotation_tasks - TypeError: and_ expected 2 arguments, got 5
================================================== 1 failed, 98 passed, 837 deselected in 1.16s ================================================== 

(annotation-py3.8) C:\Users\milan_bogicevic\Desktop\epam_project\badgerdoc\annotation>
```